### PR TITLE
[t5 model parallel] misc fixes

### PR DESCRIPTION
**Pull request by stas00**
This PR:

* in 2 places fixes an assumption that devices on the device map are always ` (0, 1, 2, 3)` and:
   1. are ordered by their cuda device id and not `(2, 3, 0, 1)`
   2. have a stride of 1 and not `(0, 2, 3, 5) `
* adds a missing `to()`, removes a redundant `to()`
* removes obvious comments
* removes code that gets run twice
* this PR continues at #9323 -  I branched off from this PR and implemented an automatic remap of inputs and a lot refactoring. 

I will comment on the reasons for changes in the code.

There is one gotcha wrt py36 w/o cython not having its dict ordered. Please see https://github.com/huggingface/transformers/pull/9316#discussion_r549068073

I think sorting out the logic first device/last device/is_this_the_last_layer_of_this_device and such logic should be abstracted away for readability, and not needing to replicate the same logic in each model. Perhaps `self.device_map` should be a smart class that can provide all the answers via its methods.

@alexorona, I'm studying your t5-mp implementation to do the same for bart. Thank you for doing the hard work of putting the foundation in place and porting 2 models!!!

Please have a look and let me know if my tweaks make sense. Your original code is excellent - I'm just trying to think how to make it easier to replicate it in other models and improve readability, hence a gazillion of questions/suggestions.

Also, if you don't mind I have a few design questions:

1. Could you please comment on why you are splitting the encoder between all devices on the device map and the same for the decoder? Won't it be more efficient performance-wise to put the encoder on the first group of devices and decoder on the second? 

2. I also find it confusing that the device map doesn't map out the whole model, but just the encoder and assumes that the decoder has the same config. I'm not familiar with t5 but other models definitely can have encoder and decoder that don't at all match number of layers-wise. And while this is perhaps not the case for t5, I think the device map should be intuitively similar for all models as we slowly progress with porting other models to MP. That is I think it should include all layers of the model and not half of them.

@patrickvonplaten, @LysandreJik

### CODE REVIEW

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -785,17 +785,15 @@ def parallelize(self, device_map=None):
         )
         assert_device_map(self.device_map, len(self.block))
         self.model_parallel = True
-        self.first_device = "cpu" if "cpu" in self.device_map.keys() else "cuda:" + str(min(self.device_map.keys()))
-        self.last_device = "cuda:" + str(max(self.device_map.keys()))
+        self.first_device = "cpu" if "cpu" in self.device_map.keys() else f"cuda:{ list(self.device_map.keys())[0] }"
+        self.last_device = f"cuda:{ list(self.device_map.keys())[-1] }"
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:789]: So this change deals with the case where devices aren't necessarily ordered by ID. That is the case where a device_map's keys are say, `[0,2,1,3]` for whatever reason.

The only issue here is that only starting from py37 we get ordered dicts and all is good, so for py36, unless it is a cython-built we have an issue that 0th item might not be the 0th item in the device map. Not sure how to deal with it.

Or alternatively we need to assert if the order is not numericaly rising, but again we have an issue with the dict, before py37.

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -785,17 +785,15 @@ def parallelize(self, device_map=None):
...
         # Load onto devices
         for k, v in self.device_map.items():
             for layer in v:
                 cuda_device = "cuda:" + str(k)
                 self.block[layer] = self.block[layer].to(cuda_device)
 
-        # Set embed_tokens to first layer
         self.embed_tokens = self.embed_tokens.to(self.first_device)
-        # Set final layer norm to last device
         self.final_layer_norm = self.final_layer_norm.to(self.last_device)
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:797]: these are loud and clear in the code already.

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -833,7 +831,6 @@ def forward(
         # Model parallel
         if self.model_parallel:
             torch.cuda.set_device(self.first_device)
-            self.embed_tokens = self.embed_tokens.to(self.first_device)
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:836]: this has already been called in `parallelize` - did I miss that perhaps it may have changed since `parallelize` or would it be better to call it here rather than in `parallelize`?

File: src/transformers/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -951,9 +948,10 @@ def forward(
 
             # Model Parallel: If it's the last layer for that device, put things on the next device
             if self.model_parallel:
-                for k, v in self.device_map.items():
-                    if i == v[-1] and "cuda:" + str(k) != self.last_device:
-                        hidden_states = hidden_states.to("cuda:" + str(k + 1))
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:956]: and here we have an issue of an assumption that devices are not only ordered but also have a stride of 1, so the above would fail if we have `[0,2,1,3]`.

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:797]: these are loud and clear in the code already.

Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:797]: these are loud and clear in the code already.

File: src/transformers/models/t5/modeling_t5.py (hunk 4 above)
Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: the code is very hard to parse, that's why I think all these should be abstracted away, e.g. we should be able to say:
```
device_map.next_device(current_device)
```

File: src/transformers/models/t5/modeling_t5.py [hunk 5]
```diff
@@ -904,9 +901,7 @@ def forward(
             # Model parallel
             if self.model_parallel:
                 torch.cuda.set_device(hidden_states.device)
-                # Ensure that attention_mask is always on the same device as hidden_states
-                if attention_mask is not None:
-                    attention_mask = attention_mask.to(hidden_states.device)
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:909]: this variable is not needed to be switched to other devices

File: src/transformers/models/t5/modeling_t5.py [hunk 6]
```diff
@@ -915,6 +910,9 @@ def forward(
                     encoder_extended_attention_mask = encoder_extended_attention_mask.to(hidden_states.device)
                 if encoder_decoder_position_bias is not None:
                     encoder_decoder_position_bias = encoder_decoder_position_bias.to(hidden_states.device)
+                if all_hidden_states is not None:
+                    all_hidden_states = all_hidden_states.to(hidden_states.device)
+
```
Comment (stas00) [inline src/transformers/models/t5/modeling_t5.py:915]: This on the other hand needs to be switched. Otherwise we end up with the next line where 2 vars could be on 2 different devices.

Moreover it's telling me, there is a test missing for this case.

File: src/transformers/models/t5/modeling_t5.py (hunk 4 above)
Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: This work:
```
            def next_device(device_map, current_layer_number):
                devices = list(device_map.keys())
                for e, d in enumerate(devices):
                    if current_layer_number == self.device_map[d][-1]:
                        if e+1 != len(devices):
                            return devices[e+1]
                        else:
                            return devices[e]
                    
            # Model Parallel: If it's the last layer for that device, put things on the next device
            if self.model_parallel:
                hidden_states = hidden_states.to(next_device(self.device_map, i))
```

Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: This perhaps is more readable/efficient:
```
            def switch_to_next_device(device_map, current_layer_number):
                """ returns False if shouldn't switch, otherwise device id to switch to """
                devices = list(device_map.keys())
                for e, d in enumerate(devices):
                    if current_layer_number == device_map[d][-1] and e+1 != len(devices):
                        return devices[e+1] # next device
                return False
                    
            # Model Parallel: If it's the last layer for that device, put things on the next device
            if self.model_parallel:
                next_device_id = switch_to_next_device(self.device_map, i)
                if next_device_id is not False:
                    hidden_states = hidden_states.to(next_device_id)
```

Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: The other even simpler approach would be to create 2 reverse dicts for:

1. layer number -> device_id
2. layer number -> next_device_id 

then it's just a one trivial lookup at the point where it's needed, with the special case of the last group of layers pointing to None for next_device_id

Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: and if we stick to `cuda:X` we should do all that remapping at the beginning too and not have the mess of adding `cuda:` everywhere. And it's easier to compare foo.device too that way.

**Comment by stas00** [comment]
Somehow I'm feeling that this approach of having special logic to remap inputs ahead of time is over-complicated. I haven't tried it yet, but won't it be much much simpler to remap inputs once the model layer is visible and just before they are used by the model - i.e. at the point where one gets:
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:1 and cuda:0!
```
Then we just do `input_foo.to(next(self.parameters()).device)` and we are done?

File: src/transformers/models/t5/modeling_t5.py (hunk 4 above)
Comment (stas00) [reply src/transformers/models/t5/modeling_t5.py:956]: Well, if #9323 is welcomed, this whole discussion while still valid becomes moot, since none of this code is needed at all.

**Comment by patrickvonplaten** [comment]
I also think #9323 is the way to go here

**Comment by stas00** [comment]
too long. closing.
